File: lib/alchemy/pillars.ts
/**
 * The Fourteen Alchemical Pillars — the isomorphic TypeScript view of
 * alchm-astro-core/spec/pillars.v1.json.
 *
 * The same JSON drives the Rust crate (alchm-astro-core/src/pillars.rs, vendored
 * into Pentacles) and the reference harness (alchm-astro-core/tools/pillar_sim.py),
 * so edit the spec, never a copy. Dependency-free: safe to import from client
 * components and from the server.
 *
 * ESMS arrays are [Spirit, Essence, Matter, Substance]; signs are 0 = Aries ..
 * 11 = Pisces; planets are 0 = Sun .. 9 = Pluto.
 */

import specJson from '@/alchm-astro-core/spec/pillars.v1.json'

export type Esms = [number, number, number, number]
export type ElementName = 'Fire' | 'Earth' | 'Air' | 'Water'
export type ModalityName = 'Cardinal' | 'Fixed' | 'Mutable'
export type Sect = 'diurnal' | 'nocturnal' | 'both'
/** The sky a caster stands under at cast time. */
export type Sky = Exclude<Sect, 'both'>
export type CastMode = 'self' | 'target'
export type PlanetName =
  | 'Sun'
  | 'Moon'
  | 'Mercury'
  | 'Venus'
  | 'Mars'
  | 'Jupiter'
  | 'Saturn'
  | 'Uranus'
  | 'Neptune'
  | 'Pluto'

export type PillarKey =
  | 'Solution'
  | 'Filtration'
  | 'Evaporation'
  | 'Distillation'
  | 'Separation'
  | 'Rectification'
  | 'Calcination'
  | 'Comixion'
  | 'Purification'
  | 'Inhibition'
  | 'Fermentation'
  | 'Fixation'
  | 'Multiplication'
  | 'Protection'

export interface PillarSpec {
  id: number
  key: PillarKey
  name: string
  effects: Esms
  primary: ElementName
  secondary: ElementName | null
  sect: Sect
  castMode: CastMode
  rulers: PlanetName[]
  /** Balance coefficient, tuned with tools/pillar_sim.py balance. */
  k: number
  /** Protection: unlocks only when the Sun or Moon has positive dignity. */
  requiresLuminaryDignity: boolean
}

export interface PillarConstants {
  poolBaseline: number
  castCharge: number
  powerRef: number
  magnitudeMin: number
  magnitudeMax: number
  deltaScale: number
  minPrimaryPlacements: number
  minHand: number
  roomResistanceFloor: number
  epsilon: number
  tieTolerance: number
}

/** The minimum a chart needs for pillar math. */
export interface ChartInput {
  /** Sign (0..11) of each planet, Sun..Pluto. */
  signs: readonly number[]
  ascendantSign: number
  /** A time-unknown chart's Ascendant is a noon placeholder, so it is not counted. */
  timeKnown: boolean
}

export const PILLAR_SPEC_VERSION: string = specJson.version
export const PILLAR_CONSTANTS: Readonly<PillarConstants> = specJson.constants
export const PILLARS: readonly PillarSpec[] = specJson.pillars as unknown as PillarSpec[]
export const PILLAR_KEYS: readonly PillarKey[] = PILLARS.map(p => p.key)

const SIGN_ORDER: readonly string[] = specJson.conventions.signOrder
const ELEMENTS: readonly ElementName[] = ['Fire', 'Earth', 'Air', 'Water']
const MODALITIES: readonly ModalityName[] = ['Cardinal', 'Fixed', 'Mutable']

/** Each planet's base ESMS contribution, indexed Sun..Pluto. */
export const PLANET_ALCHEMY: readonly Esms[] = specJson.natal.planets.map(p => p.alchemy as Esms)
export const ASCENDANT_ALCHEMY: Esms = specJson.natal.ascendantAlchemy as Esms
/** Dignity of each planet (Sun..Pluto) in each sign (Aries..Pisces). */
export const PLANET_DIGNITY: readonly (readonly number[])[] = specJson.natal.planets.map(p => {
  const table: Partial<Record<string, number>> = p.dignity
  return SIGN_ORDER.map(sign => table[sign] ?? 0)
})

export function isPillarKey(v: unknown): v is PillarKey {
  return typeof v === 'string' && PILLAR_KEYS.includes(v as PillarKey)
}

export function pillarById(id: number): PillarSpec | undefined {
  return PILLARS.find(p => p.id === id)
}

export function pillarByKey(key: PillarKey): PillarSpec {
  const p = PILLARS.find(x => x.key === key)
  if (!p) throw new Error(`Unknown pillar ${key}`)
  return p
}

export function elementOfSign(sign: number): ElementName {
  return ELEMENTS[sign % 4]
}

export function dignity(planet: number, sign: number): number {
  return PLANET_DIGNITY[planet][sign % 12]
}

export function countedSigns(chart: ChartInput): number[] {
  return chart.timeKnown ? [...chart.signs, chart.ascendantSign] : [...chart.signs]
}

export function elementCounts(chart: ChartInput): Record<ElementName, number> {
  const counts: Record<ElementName, number> = { Fire: 0, Earth: 0, Air: 0, Water: 0 }
  for (const sign of countedSigns(chart)) counts[ELEMENTS[sign % 4]] += 1
  return counts
}

export function modalityCounts(chart: ChartInput): Record<ModalityName, number> {
  const counts: Record<ModalityName, number> = { Cardinal: 0, Fixed: 0, Mutable: 0 }
  for (const sign of countedSigns(chart)) counts[MODALITIES[sign % 3]] += 1
  return counts
}

/**
 * Pillar ids (ascending) a chart may cast under `sky`: sect matches, at least
 * `minPrimaryPlacements` placements in the primary element, topped up to
 * `minHand` from the caster's strongest elements.
 */
export function hand(chart: ChartInput, sky: Sky): number[] {
  const counts = elementCounts(chart)
  const luminary = dignity(0, chart.signs[0]) > 0 || dignity(1, chart.signs[1]) > 0
  const eligible = (p: PillarSpec) =>
    (p.sect === sky || p.sect === 'both') && (!p.requiresLuminaryDignity || luminary)

  const ids = PILLARS.filter(
    p => eligible(p) && counts[p.primary] >= PILLAR_CONSTANTS.minPrimaryPlacements
  ).map(p => p.id)

  if (ids.length < PILLAR_CONSTANTS.minHand) {
    const rest = PILLARS.filter(p => eligible(p) && !ids.includes(p.id)).sort(
      (a, b) => counts[b.primary] - counts[a.primary] || a.id - b.id
    )
    for (const p of rest) {
      if (ids.length >= PILLAR_CONSTANTS.minHand) break
      ids.push(p.id)
    }
  }
  return ids.sort((a, b) => a - b)
}

File: alchm-astro-core/src/pillars.rs
//! The Fourteen Alchemical Pillars, chart tallies, and hand gating.
//!
//! Mirrors `spec/pillars.v1.json`; `tests/spec_parity.rs` fails if the two drift.
//! ESMS arrays are ordered [Spirit, Essence, Matter, Substance]; signs are
//! 0 = Aries .. 11 = Pisces; planets follow `Planet` (Sun = 0 .. Pluto = 9).

use crate::{sign_modality, Modality, Planet};

/// A classical element, as carried by a zodiac sign (sign index % 4).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Element {
    Fire,
    Earth,
    Air,
    Water,
}

impl Element {
    pub const ALL: [Element; 4] = [Element::Fire, Element::Earth, Element::Air, Element::Water];

    pub fn of_sign(sign: u8) -> Element {
        Self::ALL[(sign % 4) as usize]
    }

    pub fn idx(self) -> usize {
        self as usize
    }

    pub fn name(self) -> &'static str {
        match self {
            Element::Fire => "Fire",
            Element::Earth => "Earth",
            Element::Air => "Air",
            Element::Water => "Water",
        }
    }
}

/// Which sky a pillar can be cast under.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Sect {
    Diurnal,
    Nocturnal,
    Both,
}

/// Where a pillar's ESMS delta lands: on the caster, or on the opponent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CastMode {
    SelfCast,
    Target,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PillarSpec {
    pub id: u8,
    pub key: &'static str,
    /// ±1 per principle, [Spirit, Essence, Matter, Substance].
    pub effects: [i8; 4],
    pub primary: Element,
    pub secondary: Option<Element>,
    pub sect: Sect,
    pub cast_mode: CastMode,
    pub rulers: &'static [Planet],
    /// Balance coefficient, tuned with `tools/pillar_sim.py balance`.
    pub k: f64,
    /// Protection: unlocks only when the Sun or Moon has positive dignity.
    pub requires_luminary_dignity: bool,
}

use CastMode::{SelfCast, Target};
use Element::{Air, Earth, Fire, Water};
use Planet::*;
use Sect::{Both, Diurnal, Nocturnal};

pub const PILLARS: [PillarSpec; 14] = [
    PillarSpec { id: 1, key: "Solution", effects: [-1, 1, 1, -1], primary: Water, secondary: Some(Earth), sect: Nocturnal, cast_mode: Target, rulers: &[Moon, Neptune], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 2, key: "Filtration", effects: [1, 1, -1, 1], primary: Air, secondary: Some(Water), sect: Nocturnal, cast_mode: SelfCast, rulers: &[Mercury, Saturn], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 3, key: "Evaporation", effects: [1, 1, -1, -1], primary: Air, secondary: Some(Fire), sect: Diurnal, cast_mode: SelfCast, rulers: &[Mercury, Uranus], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 4, key: "Distillation", effects: [1, 1, -1, 1], primary: Water, secondary: Some(Air), sect: Diurnal, cast_mode: SelfCast, rulers: &[Mercury, Neptune], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 5, key: "Separation", effects: [1, 1, 1, -1], primary: Fire, secondary: Some(Water), sect: Diurnal, cast_mode: Target, rulers: &[Mercury, Uranus, Pluto], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 6, key: "Rectification", effects: [1, 1, 1, 1], primary: Fire, secondary: None, sect: Diurnal, cast_mode: Target, rulers: &[Sun, Jupiter], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 7, key: "Calcination", effects: [-1, 1, 1, -1], primary: Fire, secondary: Some(Earth), sect: Diurnal, cast_mode: Target, rulers: &[Mars, Saturn], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 8, key: "Comixion", effects: [1, -1, 1, 1], primary: Earth, secondary: Some(Air), sect: Nocturnal, cast_mode: Target, rulers: &[Venus, Jupiter, Pluto], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 9, key: "Purification", effects: [1, 1, -1, -1], primary: Fire, secondary: Some(Air), sect: Diurnal, cast_mode: SelfCast, rulers: &[Mercury, Neptune, Moon], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 10, key: "Inhibition", effects: [-1, -1, 1, 1], primary: Earth, secondary: Some(Water), sect: Nocturnal, cast_mode: Target, rulers: &[Saturn, Pluto], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 11, key: "Fermentation", effects: [1, 1, 1, -1], primary: Water, secondary: Some(Fire), sect: Nocturnal, cast_mode: Target, rulers: &[Pluto, Jupiter, Mars], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 12, key: "Fixation", effects: [-1, -1, 1, 1], primary: Earth, secondary: Some(Air), sect: Diurnal, cast_mode: Target, rulers: &[Saturn, Venus], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 13, key: "Multiplication", effects: [1, 1, 1, -1], primary: Fire, secondary: Some(Water), sect: Nocturnal, cast_mode: Target, rulers: &[Jupiter, Sun, Uranus], k: 1.0, requires_luminary_dignity: false },
    PillarSpec { id: 14, key: "Protection", effects: [1, 1, 1, 1], primary: Fire, secondary: Some(Earth), sect: Both, cast_mode: Target, rulers: &[Sun, Moon, Mercury, Jupiter], k: 1.0, requires_luminary_dignity: true },
];

pub fn pillar(id: u8) -> Option<&'static PillarSpec> {
    PILLARS.iter().find(|p| p.id == id)
}

// ── Natal tables (ASOL lib/alchemizer.ts) ───────────────────────────────────

/// Each planet's base ESMS contribution, [Spirit, Essence, Matter, Substance].
pub const PLANET_ALCHEMY: [[f64; 4]; 10] = [
    [1.0, 0.0, 0.0, 0.0], // Sun
    [0.0, 1.0, 1.0, 0.0], // Moon
    [1.0, 0.0, 0.0, 1.0], // Mercury
    [0.0, 1.0, 1.0, 0.0], // Venus
    [0.0, 1.0, 1.0, 0.0], // Mars
    [1.0, 1.0, 0.0, 0.0], // Jupiter
    [1.0, 0.0, 1.0, 0.0], // Saturn
    [0.0, 1.0, 1.0, 0.0], // Uranus
    [0.0, 1.0, 0.0, 1.0], // Neptune
    [0.0, 1.0, 1.0, 0.0], // Pluto
];

/// The Ascendant counts toward element and modality tallies but carries no ESMS.
pub const ASCENDANT_ALCHEMY: [f64; 4] = [0.0, 0.0, 0.0, 0.0];

/// Dignity of each planet in each sign (Aries..Pisces); 0 where unlisted.
pub const PLANET_DIGNITY: [[i8; 12]; 10] = [
    [2, 0, 0, 0, 1, 0, -2, 0, 0, 0, -1, 0],   // Sun
    [0, 2, 0, 1, 0, 0, 0, -2, 0, -1, 0, 0],   // Moon
    [0, 0, 1, 0, 0, 3, 0, 0, 1, 0, 0, -3],    // Mercury
    [-1, 1, 0, 0, 0, -2, 1, -1, 0, 0, 0, 2],  // Venus
    [1, -1, 0, -2, 0, 0, -1, 1, 0, 2, 0, 0],  // Mars
    [0, 0, -1, 2, 0, -1, 0, 0, 1, -2, 0, 1],  // Jupiter
    [-2, 0, 0, -1, -1, 0, 2, 0, 0, 1, 1, 0],  // Saturn
    [0, -3, 0, 0, 0, 0, 0, 2, 0, 0, 1, 0],    // Uranus
    [0, 0, 0, 2, 0, -1, 0, 0, 0, -2, 0, 1],   // Neptune
    [0, -1, 0, 0, 2, 0, 0, 1, 0, 0, -2, 0],   // Pluto
];

pub fn dignity(planet: Planet, sign: u8) -> i8 {
    PLANET_DIGNITY[planet.idx()][(sign % 12) as usize]
}

// ── Chart tallies ───────────────────────────────────────────────────────────

/// The minimum a chart needs for pillar math: ten planet signs and an Ascendant.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ChartInput {
    /// Sign of each planet, indexed by `Planet`.
    pub signs: [u8; 10],
    pub ascendant_sign: u8,
    /// A time-unknown chart's Ascendant is a noon placeholder, so it is not counted.
    pub time_known: bool,
}

impl ChartInput {
    pub fn counted_signs(&self) -> impl Iterator<Item = u8> + '_ {
        self.signs.iter().copied().chain(self.time_known.then_some(self.ascendant_sign))
    }

    pub fn bodies(&self) -> usize {
        if self.time_known { 11 } else { 10 }
    }
}

/// Placements per element, indexed by `Element`.
pub fn element_counts(chart: &ChartInput) -> [u32; 4] {
    let mut counts = [0u32; 4];
    for sign in chart.counted_signs() {
        counts[Element::of_sign(sign).idx()] += 1;
    }
    counts
}

/// Placements per modality, [Cardinal, Fixed, Mutable].
pub fn modality_counts(chart: &ChartInput) -> [u32; 3] {
    let mut counts = [0u32; 3];
    for sign in chart.counted_signs() {
        let i = match sign_modality(sign) {
            Modality::Cardinal => 0,
            Modality::Fixed => 1,
            Modality::Mutable => 2,
        };
        counts[i] += 1;
    }
    counts
}

// ── Hand ────────────────────────────────────────────────────────────────────

/// A pillar needs this many placements in its primary element to unlock.
pub const MIN_PRIMARY_PLACEMENTS: u32 = 2;
/// Every caster holds at least this many pillars, topped up from their strongest element.
pub const MIN_HAND: usize = 2;

/// Pillar ids (ascending) the chart may cast under `sky` (Diurnal or Nocturnal).
pub fn hand(chart: &ChartInput, sky: Sect) -> Vec<u8> {
    let counts = element_counts(chart);
    let luminary = dignity(Sun, chart.signs[Sun.idx()]) > 0 || dignity(Moon, chart.signs[Moon.idx()]) > 0;
    let eligible = |p: &PillarSpec| {
        (p.sect == sky || p.sect == Both) && (!p.requires_luminary_dignity || luminary)
    };

    let mut ids: Vec<u8> = PILLARS
        .iter()
        .filter(|p| eligible(p) && counts[p.primary.idx()] >= MIN_PRIMARY_PLACEMENTS)
        .map(|p| p.id)
        .collect();

    if ids.len() < MIN_HAND {
        let mut rest: Vec<&PillarSpec> = PILLARS.iter().filter(|p| eligible(p) && !ids.contains(&p.id)).collect();
        rest.sort_by_key(|p| (std::cmp::Reverse(counts[p.primary.idx()]), p.id));
        for p in rest {
            if ids.len() >= MIN_HAND {
                break;
            }
            ids.push(p.id);
        }
    }
    ids.sort_unstable();
    ids
}
